lib/SILGen/SILGenExpr.cpp

```diff
@@ -3648,6 +3682,11 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
     rootTy = rootTy->mapTypeOutOfContext()->getCanonicalType();
   }
   
+  // Treat a dynamic Self as the statically-known type.
```

Comment:
This is not correct, because dynamic Self needs to be the actual dynamic type of the object the method was invoked on. If you have:

```
class Base {
  var foo: Self { return self }
  func bar() -> AnyKeyPath { \Self.foo }
}

class Derived: Base {}

Derived().bar()
```

then the result should be `\Derived.foo`, not `\Base.foo`. In order to do this in full generality, we need to emit dynamic Self as a generic argument to the key path pattern, something like:

```
  %kp = keypath $KeyPath<Self, Self>, <T: Base> (root $T; gettable_property $T, ...) <Self>
```